File: src/main/java/com/aws/proseve/ggcommons/config/manager/GreengrassConfigManager.java
package com.aws.proseve.ggcommons.config.manager;

import com.github.cliftonlabs.json_simple.JsonObject;
import com.github.cliftonlabs.json_simple.Jsoner;
import software.amazon.awssdk.aws.greengrass.GreengrassCoreIPCClientV2;
import software.amazon.awssdk.aws.greengrass.model.GetConfigurationRequest;
import software.amazon.awssdk.aws.greengrass.model.GetConfigurationResponse;

import java.util.Map;

public class GreengrassConfigManager extends ConfigManager
{
    String configComponentName;
    String configKey;

    GreengrassConfigManager(String componentName, String configComponentName, String configKey)
    {
        super(componentName);
        this.configComponentName = configComponentName;
        this.configKey = (configKey == null) ? "ComponentConfig" : configKey;
        init();
    }

    @Override
    protected JsonObject loadConfiguration()
    {
        JsonObject retVal = null;
        LOGGER.debug("Loading Greengrass component configuration");

        try (GreengrassCoreIPCClientV2 ipcClient = connectToIPC())
        {
            GetConfigurationRequest request;
            if (configComponentName == null) {
                request = new GetConfigurationRequest();
            } else
            {
                request = new GetConfigurationRequest().withComponentName(configComponentName);
            }
            GetConfigurationResponse response = ipcClient.getConfiguration(request);
            Map<String,Object> responseValue = response.getValue();
            if (responseValue != null)
            {
                JsonObject tempConfig = new JsonObject(response.getValue());
                JsonObject fullConfig = (JsonObject) Jsoner.deserialize(tempConfig.toJson());
                LOGGER.info("Full configuration retrieved from Nucleus: {}", fullConfig.toJson());
                retVal = (JsonObject) fullConfig.get(configKey);
                LOGGER.info("Component configuration retrieved from Nucleus: {}", retVal.toJson());
            } else {
                LOGGER.fatal("Configuration not found.  Incorrect component name?  Exiting");
                System.exit(5);
            }
        }
        catch (Exception e)
        {
            throw new RuntimeException(e);
        }

        return retVal;
    }

    @Override
    protected String getConfigSource()
    {
        return String.format("Greengrass com.aws.proseve.ggcommons.config (component: %s; key: %s)", configComponentName, configKey);
    }

    private GreengrassCoreIPCClientV2 connectToIPC()
    {
        GreengrassCoreIPCClientV2 ipcClient = null;
        try
        {
            ipcClient = GreengrassCoreIPCClientV2.builder().build();
        }
        catch (Exception e)
        {
            LOGGER.fatal("Unable to connect to Greengrass IPC to retrieve configuration.");
            System.exit(5);
        }
        return ipcClient;
    }
}

File: libs/python/edgecommons/config/manager/greengrass_config_manager.py
import logging

from awsiot.greengrasscoreipc.clientv2 import GreengrassCoreIPCClientV2

from edgecommons.config.manager.config_manager import ConfigManager

logger = logging.getLogger("GreengrassConfigManager")


class GreengrassConfigManager(ConfigManager):
    def __init__(
        self,
        thing_name: str,
        component_name: str,
        config_component_name: str,
        config_key: str,
        platform=None,
        candidate_validators=None,
        validation_timeout_secs=5.0,
    ):
        super().__init__(
            component_name,
            thing_name,
            platform=platform,
            candidate_validators=candidate_validators,
            validation_timeout_secs=validation_timeout_secs,
        )
        self._config_component_name = config_component_name
        self._config_key = config_key if config_key is not None else "ComponentConfig"
        self._config_source = f"Greengrass config (component: {config_component_name}; key: {self._config_key})"
        self._config_provider_family = "GG_CONFIG"
        self.init()

    def _load_configuration(self) -> dict:
        logger.info(
            f"Loading Greengrass component configuration from component '{self._config_component_name}'"
        )
        ipc_client = GreengrassCoreIPCClientV2()
        if self._config_component_name is None:
            response = ipc_client.get_configuration()
        else:
            response = ipc_client.get_configuration(
                component_name=self._config_component_name
            )
        logger.debug(f"Full configuration retrieved from Nucleus: {response.value}")
        ret_val = None
        if response.value is not None:
            if self._config_key in response.value:
                ret_val = response.value.get(self._config_key)
                logger.debug(f"Component configuration retrieved: {ret_val}")
            else:
                ipc_client.close()
                raise RuntimeError(
                    f"Configuration not found in component '{self._config_component_name}' at key '{self._config_key}'"
                )
        ipc_client.close()
        return ret_val

